src/top/run/box_generator.py

```diff
@@ -0,0 +1,227 @@
+"""
+General 3D Bounding Box class & various method for the box.
+Reference: https://github.com/google-research-datasets/Objectron/blob/master/objectron/dataset/box.py
+"""
+
+
+import numpy as np
+from numpy.linalg import lstsq
+from scipy.spatial.transform import Rotation
+
+EDGES = ([1, 5], [2, 6], [3, 7], [4, 8],  # lines along x-axis
+         [1, 3], [5, 7], [2, 4], [6, 8],  # lines along y-axis
+         [1, 2], [3, 4], [5, 6], [7, 8])  # lines along z-axis
+
+
+# The vertices are ordered according to the left-hand rule,
+# so the normal vector of each face will point inward the box.
+FACES = np.array([[5, 6, 8, 7],     # +x on yz plane
+                  [1, 3, 4, 2],     # -x on yz plane
+                  [3, 7, 8, 4],     # +y on xz plane = top
+                  [1, 2, 6, 5],     # -y on xz plane
+                  [2, 4, 8, 6],     # +z on xy plane = front
+                  [1, 5, 7, 3],])   # -z on xy plane
+
+UNIT_BOX = np.asarray([[0., 0., 0.],        # center
+                       [-0.5, -0.5, -0.5],
+                       [-0.5, -0.5, 0.5],
+                       [-0.5, 0.5, -0.5],
+                       [-0.5, 0.5, 0.5],
+                       [0.5, -0.5, -0.5],
+                       [0.5, -0.5, 0.5],
+                       [0.5, 0.5, -0.5],
+                       [0.5, 0.5, 0.5],])
+
+NUM_KEYPOINTS = 8+1
+FRONT_FACE_ID = 4
+TOP_FACE_ID = 2
+
+
+class Box(object):
+    """General 3D Oriented Bounding Box."""
+
+    def __init__(self, vertices=None):
+        if vertices is None:
+            vertices = self.scaled_axis_aligned_vertices(np.array([1., 1., 1.]))
+
+        self._vertices = vertices
+        self._rotation = None
+        self._translation = None
+        self._scale = None
+        self._transformation = None
+        self._volume = None
+
+    @classmethod
+    def from_transformation(cls, rotation, translation, scale):
+        """Constructs an oriented bounding box from transformation and scale."""
+        if rotation.size != 3 and rotation.size != 9:
+            raise ValueError('Unsupported rotation, only 3x1 euler angles or 3x3 '
+                             + 'rotation matrices are supported. ' + rotation)
+        if rotation.size == 3:
+            rotation = Rotation.from_rotvec(rotation.tolist()).as_dcm()
+        scaled_identity_box = cls.scaled_axis_aligned_vertices(scale)
+        vertices = np.zeros((NUM_KEYPOINTS, 3))
+        for i in range(NUM_KEYPOINTS):
```

Comment:
For loop is not needed here if you think about which components of the tensor interact with each other.
We have a rotation matrix of shape (3,3) and an input list of vertices (N,3) that is transformed to the output list of vertices (N,3).